src/main/java/net/hydromatic/lookml/Values.java

```diff
@@ -0,0 +1,199 @@
+  static class NumberValue extends Value {
+    private final Number number;
+
+    NumberValue(Number number) {
+      this.number = number;
+    }
+
+    @Override void write(LookmlWriter writer) {
+      writer.number(number);
+    }
+  }
+
+  /** Value of a property or list element whose value is a string. */
+  static class StringValue extends Value {
+    final String s;
+
+    StringValue(String s) {
+      this.s = s;
+    }
+
+    @Override void write(LookmlWriter writer) {
+      writer.string(s);
+    }
+  }
+
+  /** Value of a property whose value is a code block. */
+  static class CodeValue extends Value {
+    private final String s;
+
+    CodeValue(String s) {
+      this.s = s;
+    }
+
+    @Override void write(LookmlWriter writer) {
+      writer.code(s);
+    }
+  }
+
+  /** Value of a property whose value is a ref-string pair. */
+  static class PairValue extends Value {
+    final String ref;
+    final String s;
+
+    PairValue(String ref, String s) {
+      this.ref = ref;
+      this.s = s;
+    }
+
+    @Override void write(LookmlWriter writer) {
+      writer.label(ref);
+      writer.string(s);
+    }
+  }
+
+  /** Value of a property or list element whose value is a list. */
+  static class ListValue extends Value {
+    final List<Value> list;
+
+    ListValue(List<Value> list) {
+      this.list = list;
+    }
+
+    @Override void write(LookmlWriter writer) {
+      writer.list(list);
+    }
+  }
+
+  /** Value of a property whose value is an object.
+   *
+   * <p>For example,
+   * <blockquote><pre>{@code
+   * conditionally_filter: {
+   *   filters: [f3: "> 10"]
+   *   unless: [f1, f2]
+   * }
+   * }</pre></blockquote>
+   *
+   * <p>The name of the property, {@code conditionally_filter}, will be held
+   * in the enclosing property list. */
+  static class ObjectValue extends Value {
+    final ImmutablePairList<String, Value> properties;
+
+    ObjectValue(PairList<String, Value> properties) {
+      this.properties = ImmutablePairList.copyOf(properties);
+    }
+
+    @Override void write(LookmlWriter writer) {
+      writer.obj(properties);
+    }
+  }
+
+  /** Value of a property whose value is an object and that also has a name.
+   *
+   * <p>For example,
+   * <blockquote><pre>{@code
+   * dimension: d1 {
+   *   sql: orderDate;;
+   *   type: int
+   * }
+   * }</pre></blockquote>
+   *
+   * <p>{@link #name} is "d1", and {@link #properties} has entries for "sql" and
+   * "type". The name of the property, {@code dimension}, is held in the
+   * enclosing property list. */
+  static class NamedObjectValue extends ObjectValue {
+    final String name;
+
+    NamedObjectValue(String name, PairList<String, Value> properties) {
+      super(properties);
+      this.name = requireNonNull(name);
+    }
+
+    @Override void write(LookmlWriter writer) {
+      // Write "name { ... }"
+      writer.identifier(name);
+      writer.buf.append(' ');
```

Comment:
Why does this write directly to the buffer in the writer?  Shouldn't the writer handle all output logic/direct access to its buffer?

Reply:
Fixed